zvariant/src/file_path.rs

```diff
@@ -0,0 +1,293 @@
+    }
+}
+
+impl<'f> From<&'f OsString> for FilePath<'f> {
+    fn from(value: &'f OsString) -> FilePath<'f> {
+        FilePath(bytes_with_null(value.as_encoded_bytes()))
+    }
+}
+
+impl From<OsString> for FilePath<'_> {
+    fn from(value: OsString) -> Self {
+        FilePath(vec_to_cstr(value.as_encoded_bytes().to_vec()))
+    }
+}
+
+impl<'f> From<&'f PathBuf> for FilePath<'f> {
+    fn from(value: &'f PathBuf) -> FilePath<'f> {
+        FilePath::from(value.as_os_str())
+    }
+}
+
+impl<'f> From<PathBuf> for FilePath<'f> {
+    fn from(value: PathBuf) -> FilePath<'f> {
+        FilePath::from(OsString::from(value))
+    }
+}
+
+impl<'f> From<&'f Path> for FilePath<'f> {
+    fn from(value: &'f Path) -> Self {
+        Self::from(value.as_os_str())
+    }
+}
+
+impl<'f> From<&'f CStr> for FilePath<'f> {
+    fn from(value: &'f CStr) -> Self {
+        Self(Cow::Borrowed(value))
+    }
+}
+
+impl<'f> From<&'f str> for FilePath<'f> {
+    fn from(value: &'f str) -> Self {
+        Self::from(OsStr::new(value))
+    }
+}
+
+impl<'f> AsRef<FilePath<'f>> for FilePath<'f> {
+    fn as_ref(&self) -> &FilePath<'f> {
+        self
+    }
+}
+
+impl<'f> From<FilePath<'f>> for OsString {
+    fn from(value: FilePath<'f>) -> Self {
+        unsafe { OsString::from_encoded_bytes_unchecked(value.0.to_bytes().to_vec()) }
+    }
+}
+
+impl<'f> From<&'f FilePath<'f>> for &'f Path {
+    fn from(value: &'f FilePath<'f>) -> Self {
+        // unwrap?
+        Path::new(value.0.as_ref().to_str().unwrap())
+    }
+}
+
+impl<'f> From<FilePath<'f>> for PathBuf {
+    fn from(value: FilePath<'f>) -> Self {
+        PathBuf::from(value.0.to_string_lossy().to_string())
+    }
+}
+
+/// Converts a `Vec<u8>` into a null-terminated `CStr`.
+///
+/// Truncates the vector at the first null byte, if present.
+/// If no null byte exists, appends one to ensure proper termination.
+///
+/// # Returns
+///
+/// A `Cow<'_, CStr>` containing a *guaranteed* null-terminated string.
+pub fn vec_to_cstr<'a>(mut bytes: Vec<u8>) -> Cow<'a, CStr> {
+    if let Some(pos) = bytes.iter().position(|&b| b == 0) {
+        bytes.truncate(pos + 1);
+    } else {
+        bytes.push(0);
+    }
+    // SAFETY: unwrap is fine here since we the null termination
+    Cow::Owned(CString::from_vec_with_nul(bytes).unwrap())
+}
+
+/// Converts a byte slice into a null-terminated [CStr].
+///
+/// Returns a borrowed [CStr] if the slice already contains a null byte;
+/// otherwise, returns an owned [CStr] with a null byte appended.
+///
+/// # Returns
+///
+/// A [Cow<'_, CStr>] containing a *guaranteed* null-terminated string.
+fn bytes_with_null(bytes: &[u8]) -> Cow<'_, CStr> {
+    // SAFETY: unwrap is fine here since in bot cases we guarante the null termination
+    if let Some(pos) = bytes.iter().position(|&b| b == 0) {
+        return Cow::Borrowed(CStr::from_bytes_with_nul(&bytes[..=pos]).unwrap());
+    }
+    Cow::Owned(CString::new(bytes).unwrap())
+}
```

Comment:
Using [`from_bytes_until_nul`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_until_nul) one can get rid of the first unwrap.


```suggestion
fn bytes_with_null(bytes: &[u8]) -> Cow<'_, CStr> {
    if let Ok(cstr) = CStr::from_bytes_until_nul(&bytes) {
        return Cow::Borrowed(cstr);
    }
    // unwrap is fine, as we handled the null containing case above
    Cow::Owned(CString::new(bytes).unwrap())
}
```

Reply:
Make sense, I'll apply that and force push.